python/triton/compiler/code_generator.py

```diff
@@ -1227,16 +1264,32 @@ def call_JitFunction(self, fn: JITFunction, args, kwargs):
                 raise CompilationError(self.jit_fn.src, self.cur_node, None) from e
 
             callee_ret_type = generator.ret_type
-            self.function_ret_types[fn_name] = callee_ret_type
+            args_mut = generator.prototype.args_mut
+            self.function_ret_types[fn_name] = (callee_ret_type, args_mut)
         else:
-            callee_ret_type = self.function_ret_types[fn_name]
+            callee_ret_type, args_mut = self.function_ret_types[fn_name]
         symbol = self.module.get_function(fn_name)
         args_val = flatten_values_to_ir(args_val)
         call_op = self.builder.call(symbol, args_val)
-        if callee_ret_type == language.void:
-            return None
         handles = [call_op.get_result(i) for i in range(call_op.get_num_results())]
-        return next(unflatten_ir_values(handles, [callee_ret_type]))
+
+        all_ret_types = ([callee_ret_type]
+                         if callee_ret_type != language.void else []) + [ty for (name, ty) in args_mut]
+        all_ret_values = unflatten_ir_values(handles, all_ret_types)
+        normal_ret = next(all_ret_values) if callee_ret_type != language.void else None
+        # Unpack and emit the writeback for mutable arguments.
+        for name, ty in args_mut:
+            target, value = getattr(call_args[name], "__ast__", None), next(all_ret_values)
+            # Discard mutation on an RValue.
+            if target is None or isinstance(target, ast.Call):
+                continue
+            if isinstance(target, ast.keyword):
+                target = target.value
+            if not isinstance(target, ValueDest):
+                target = copy.deepcopy(target)
+            target.ctx = ast.Store()
+            self.assignTarget(target, value)
```

Comment:
What if there are multiple references to the same value, e.g.
```python
a = Aggregate(...)
b = a
f(b) # a and b should be updated
```

Reply:
This will only update `b`. I might be able to fix this one.